### PR TITLE
COORD: Add a new unexpected-type error to improve messages

### PR DESCRIPTION
**Pull request by bkirwi**
In #13682, we spotted an incorrect error message -- the error said a
particular name did not exist, where in fact it did but just had an
unexpected type. Add a new error type to represent this, convert
a couple methods that ought to be using it, and remove the old type.

### Motivation

This doesn't fix the core issue in #13682, but ought to improve the user's experience in general and make similar things easier to diagnose.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

(This is technically user-visible, but I suspect does not require a release note. Happy to add one on request!)

### CODE REVIEW

**Comment by bkirwi** [comment]
@jkosh44 - following through on the error message change suggested in #13682, but let me know if you'd rather not review this one!

File: src/sql/src/catalog.rs [hunk 1]
```diff
@@ -570,6 +569,9 @@ impl fmt::Display for CatalogError {
                 write!(f, "unknown cluster replica '{}'", name)
             }
             Self::UnknownItem(name) => write!(f, "unknown catalog item '{}'", name),
+            Self::UnexpectedType(name, cit) => {
```
Comment (jkosh44) [inline src/sql/src/catalog.rs:572]: Somewhat subjective and not a big deal, but I think something like `item_type` would be better. It took me a second to figure out what `cit` meant. Feel free to ignore if you disagree.

**Review by jkosh44 — APPROVED** [review]
LGTM, but I think we should add some tests for this before merging.

File: src/sql/src/catalog.rs (hunk 1 above)
Comment (bkirwi) [reply src/sql/src/catalog.rs:572]: Fine by me!

**Comment by bkirwi** [comment]
> I think we should add some tests for this before merging. Specifically at least one where we run into this error and at least one where we can circumvent the error by using the schema name.

The former test makes sense here -- I notice some similar tests that look for the error message content in `catalog.td` -- but the second one surprises me: this PR doesn't change anything about how such a test would behave, since it doesn't actually change anything about the name resolution or search logic. Maybe the way to put it is that this PR covers only one of the two issues in #13682 (a misleading error message) and the test you propose belongs more properly to the other problem (the search path not behaving as expected).

Does that make sense? (If you feel strongly the test belongs here, I'm up for it; just checking to help improve my understanding.)

**Comment by bkirwi** [comment]
Or is the idea that the search path is fine and the only real bug was the error message, so we might as well take care of everything here?

**Comment by bkirwi** [comment]
We chatted about this live -- it's likely that other than the weird error message this is "behaving as expected," but I'll cut another PR for some tests that capture the search path behaviour more explicitly.

**Pull request merged**